packages/core/lib/use-puck-history.ts

```diff
@@ -0,0 +1,71 @@
+import type { Data } from "../types/Config";
+import { useEffect } from "react";
+import { PuckAction } from "./reducer";
+import { useActionHistory } from "./use-action-history";
+import { useHotkeys } from "react-hotkeys-hook";
+import EventEmitter from "event-emitter";
+import { useDebouncedCallback, useDebounce } from "use-debounce";
+import { diff, applyChange, revertChange } from "deep-diff";
+
+const DEBOUNCE_TIME = 100;
+const RECORD_DIFF = "RECORD_DIFF";
+export const historyEmitter = EventEmitter();
+export const recordDiff = (newData: Data) =>
+  historyEmitter.emit(RECORD_DIFF, newData);
+
+export function usePuckHistory({
+  data,
+  dispatch,
+}: {
+  data: Data;
+  dispatch: (action: PuckAction) => void;
+}) {
+  const { canForward, canRewind, rewind, forward, record } = useActionHistory();
+
+  useHotkeys("ctrl+z", rewind, [rewind]);
+  useHotkeys("ctrl+y", forward, [forward]);
+
+  const [snapshot] = useDebounce(data, DEBOUNCE_TIME);
+
+  const handleRecordDiff = useDebouncedCallback((newData: Data) => {
```

Comment:
This is excellent. What do you think about adding tests here?

No worries if not - we don't have them anywhere else yet.

Reply:
When #89 or #90 is completed, I will try with refer to that!

I think this area would be meaningful if it were basically at the level of mocking the reducer or integration testing of the puck component, but I need a better understanding of your testing philosophy.

If so, I can also help write tests for other components (or even e2e with demo app).

Reply:
That makes sense!

#89 will be unit tests, #90 more integration but I'm unclear on the strategy yet.

I'm working on #89 next. We'll pick this up afterwards.